relay-server/src/actors/outcome_aggregator.rs

```diff
@@ -0,0 +1,167 @@
+//! This module contains the outcomes aggregator, which collects similar outcomes into groups
+//! and flushed them periodically.
+
+use std::{
+    collections::{BTreeMap, HashMap},
+    net::IpAddr,
+};
+
+use actix::{Actor, AsyncContext, Context, Handler, Recipient, Supervised, SystemService};
+use relay_common::{DataCategory, UnixTimestamp};
+use relay_config::Config;
+use relay_general::protocol::EventId;
+use relay_quotas::Scoping;
+use relay_statsd::metric;
+use std::time::Duration;
+
+use crate::statsd::RelayTimers;
+
+use super::outcome::{Outcome, OutcomeError, TrackOutcome};
+
+/// Contains everything to construct a `TrackOutcome`, except quantity
+#[derive(Debug, PartialEq, Eq, Hash)]
+struct BucketKey {
+    /// Scoping of the request.
+    pub scoping: Scoping,
+    /// The outcome.
+    pub outcome: Outcome,
+    /// The event id.
+    pub event_id: Option<EventId>,
+    /// The client ip address.
+    pub remote_addr: Option<IpAddr>,
+    /// The event's data category.
+    pub category: DataCategory,
+}
+
+/// Aggregates outcomes into buckets, and flushes them periodically.
+/// Inspired by [`relay_metrics::Aggregator`].
+pub struct OutcomeAggregator {
+    /// The width of each aggregated bucket in seconds
+    bucket_interval: u64,
+    /// The time we should wait before flushing a bucket, in seconds
+    flush_delay: u64,
+    /// Maximum capacity of the actor's inbox
+    mailbox_size: usize,
+    /// Mapping from offset to bucket key to quantity. timestamp = offset * bucket_interval
+    buckets: BTreeMap<u64, HashMap<BucketKey, u32>>,
```

Comment:
why not put the timestamp into the bucket key and avoid this nesting and extra allocation? Together with `split_off` optimization this should make the flush quite fast as long as the timestamp is declared on top of `BucketKey` (see documentation for `derive(PartialOrd)`)